### PR TITLE
Show editions in reading log

### LINKED ISSUE #2733: Allow the reading log to log editions instead of works
The Reading Log feature lets keep track of the books you’re currently reading, have finished reading, or want to read. Currently it only logs works, not editions.

reading-log-screenshot

### Description of the solution I would like

I suggest changing this point to be able to log editions better than works (or editions and works alike).

### Stakeholders

I think this solution is more appealing to readers. Although often readers only share the title of the book they are reading (work), being an online reading log, I think it is more interesting to be able to share which particular book (edition) they are reading, have read, or want to read, not just the title of the original works.

### PR DESCRIPTION
**Pull request by benbdeitch**
Closes #2733

This PR alters how the 'already-read' and 'currently-read' bookshelves retrieve and display items. Rather than simply fetching the work data, they now retrieve the edition-specific data that is already stored in their reading log. 

Currently, the 'want-to-read' bookshelf is untouched, as users might want to know when any edition becomes available to borrow. 

### Technical

The `get_users_logged_books` function in openlibrary/core/Bookshelves was edited to provide this functionality. Based on the bookshelf_id received, it switches on `editions` variable, that is used to later define behavior. 

Both the `get_sorted_reading_log_books` and `get_filtered_reading_log_books` were altered to switch between fetching work data, or fetching work _and_ edition data, based on the 'editions' variable. As well, a new function, `link_editions_to_works` was created to process the retrieved results, and properly tie them together, so that they are correctly interpreted by the `SearchResultsWork` macro. 
### Testing

To verify, simply add a few books to each shelf of your reading log. Preferably, editions with titles that differ from the work's should be chosen, or a brief edit in `macro/SearchResultsWork` can be used to display the `doc_type` variable.  On each shelf, one should both check the shelf as a whole (which uses the 'get_sorted_reading_log_books' function, and attempt to search for an individual work (using the `get_filtered_reading_log_books` function). 
### Screenshot

image
image

(As noted, the 'solr_edition' line is not a part of the pull-request; merely an easy way to display the doc_type variable for testing.) 

 
image
image

### Stakeholders

@cdrini

### CODE REVIEW

**Comment by Watt3r** [comment]
I bump into this problem often and this would totally fix it for me. Thanks for implementing this @benbdeitch!

File: openlibrary/core/bookshelves.py [hunk 1]
```diff
@@ -404,25 +477,32 @@ def get_sorted_reading_log_books(
...
             )
 
             total_results = shelf_totals.get(bookshelf_id, 0)
             solr_docs = add_reading_log_data(reading_log_books, solr_docs)
 
+            # Attaches returned editions to works.
+            if editions:
+                solr_docs = cls.link_editions_to_works(solr_docs)
+
+            assert len(solr_docs) == len(reading_log_keys), (
```
Comment (jimchamp) [inline openlibrary/core/bookshelves.py:501]: Something about this branch has caused the My Books page to error with at least 2 of my accounts.
image

Comment (benbdeitch) [reply openlibrary/core/bookshelves.py:501]: I'll get to this immediately; I must have overlooked something regarding redirects. If you could do me a favor, are you able to send me the json contents of the reading logs of the accounts in question? Specifically, I'm interested in the work/edition keys involved.

Comment (jimchamp) [reply openlibrary/core/bookshelves.py:501]: I sent it to you via Slack.  Correction: this was only happening for one account.

Comment (benbdeitch) [reply openlibrary/core/bookshelves.py:501]: @jimchamp Sorry to add another request, but I'm having trouble reproducing this.  If you could do me a brief favor, could you try this with `"solr_docs found: %s, reading_log_keys:%s" % (len(solr_docs), len(reading_log_keys))` added to the feedback for the assertion error?

Comment (benbdeitch) [reply openlibrary/core/bookshelves.py:501]: image
Found some undesired behavior right here, but I'm not sure it's the same bug, or the root of the same bug.

Comment (jimchamp) [reply openlibrary/core/bookshelves.py:501]: Looks like there are 12 Solr docs and 6 reading log key objects.

**Comment by RayBB** [comment]
@benbdeitch how is it going with this? Do you need any help?

File: openlibrary/core/bookshelves.py [hunk 2]
```diff
@@ -346,13 +432,32 @@ def get_filtered_reading_log_books(
...
+
+            if editions:
+                matching_editions = [
+                    logged_work_edition_dict.get(work["key"])
+                    for work in solr_docs
+                    if logged_work_edition_dict.get(work["key"])
+                ]
+
+                edition_data = get_solr().get_many(
+                    [f'{key}' for key in matching_editions],
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:449]:
```suggestion
                    matching_editions,
```

File: openlibrary/core/bookshelves.py [hunk 3]
```diff
@@ -404,25 +509,35 @@ def get_sorted_reading_log_books(
                 oldb.query(query, vars=query_params)
             )
 
-            reading_log_work_keys = [
-                '/works/OL%sW' % i['work_id'] for i in reading_log_books
+            reading_log_keys = [
+                ('/works/OL%sW' % i['work_id'], '/books/OL%sM' % i['edition_id'])
+                if editions and i['edition_id']
+                else ('/works/OL%sW' % i['work_id'], None)
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:515]: If we don't actually care about the second element, we can create a single element tuple by adding `,` to the end. That prevents it from being treated as brackets/a string in this case.

```suggestion
                else ('/works/OL%sW' % i['work_id'],)
```

File: openlibrary/core/bookshelves.py [hunk 4]
```diff
@@ -404,25 +509,35 @@ def get_sorted_reading_log_books(
...
+                for i in reading_log_books
             ]
+
             solr_docs = get_solr().get_many(
-                reading_log_work_keys,
+                [key for key_pair in reading_log_keys for key in key_pair if key],
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:520]: We might be able to use the `flatten` method in our vendor/infogami/infogami/infobase/utils.py

```suggestion
                flatten(reading_log_keys),
```

File: openlibrary/core/bookshelves.py [hunk 5]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
 
+    # Iterates through a list of solr docs, and for all items with a 'logged edition'
+    # it will remove an item with the matching edition key from the list, and add it to
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
+        linked_docs: list[web.storage] = []
+        docs_dict = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            full_key = doc.get("key") or "/dummy/value"
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:213]: This should always have a key, so to more clearly communicate that let's drop the default. If in some very strange case it is missing the key, we likely want it to error!
```suggestion
            full_key = doc["key"]
```

File: openlibrary/core/bookshelves.py [hunk 6]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+            _, key, item_id = full_key.split("/")
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:214]:
```suggestion
            _, key, item_id = doc["key"].split("/")
```

File: openlibrary/core/bookshelves.py [hunk 7]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+            if key == "works":
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:215]: Since we only use these variables once, we can likely do this

```suggestion
            if doc['key'].startswith("/works/"):
```

File: openlibrary/core/bookshelves.py [hunk 8]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
         result = oldb.query(query, vars=data)
         return {i['bookshelf_id']: i['count'] for i in result} if result else {}
 
+    # Iterates through a list of solr docs, and for all items with a 'logged edition'
+    # it will remove an item with the matching edition key from the list, and add it to
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:208]:
```suggestion
    def link_editions_to_works(solr_docs):
    """
    :param solr_docs: Solr work/edition docs, augmented with reading log data
    """
```

File: openlibrary/core/bookshelves.py [hunk 9]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+        linked_docs: list[web.storage] = []
+        docs_dict = {}
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:210]:
```suggestion
        edition_key_to_work_doc = {}
```

File: openlibrary/core/bookshelves.py [hunk 10]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+        for edition in solr_docs:
+            full_key = edition.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "books":
+                index = docs_dict.get(item_id, -1)
+                if index >= 0:
+                    linked_docs[index].editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning("Error: No work found for edition %s" % full_key)
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:231]:
```suggestion
                if work_doc := docs_dict.get(full_key):
                    work_doc.editions = [edition]
                else:
                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
                    logger.warning("Error: No work found for edition %s" % full_key)
```

File: openlibrary/core/bookshelves.py [hunk 11]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+        docs_dict = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            full_key = doc.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "works":
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    docs_dict.update(
+                        {doc["logged_edition"].split("/")[2]: (len(linked_docs) - 1)}
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:219]: We should be able to have this map directly to the doc to make it a bit easier to read!

```suggestion
                        {doc["logged_edition"]: doc}
```

File: openlibrary/core/bookshelves.py [hunk 12]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+                if index >= 0:
+                    linked_docs[index].editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning("Error: No work found for edition %s" % full_key)
+        logger.info("LOGGERDICT IS %s" % docs_dict)
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:236]: Note to self: need to review this method.

**Review by cdrini — CHANGES_REQUESTED** [review]
This is looking great! A few code cleanups, and one potential error! Will take a look at the last method afterwards.

Testing:

This page seems to error: https://testing.openlibrary.org/people/bigdbag/books/already-read?page=10&debug=true

✅ sort | edition_id=null | https://testing.openlibrary.org/people/pyat/books/already-read
✅ filter | edition_id=null | https://testing.openlibrary.org/people/pyat/books/already-read?q=playground
✅ sort | large reading log | https://testing.openlibrary.org/people/bigdbag/books/already-read
✅ filter | large reading log | https://testing.openlibrary.org/people/bigdbag/books/already-read?q=harry+potter

File: openlibrary/core/bookshelves.py [hunk 13]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+        for idx, work_key in enumerate(reading_log_work_keys):
+            corresponding_solr_doc = next(
+                (doc for doc in solr_docs if doc.key == work_key), None
+            )
+
+            if not corresponding_solr_doc:
+                logger.info("failed key: %s" % work_key)
+                edition_key = work_edition_dict.get(work_key)
+                solr_resp = (
+                    run_solr_query(
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:271]: Having a solr query for every item here is less ideal performance wise since it means the requests the page makes is O(n) relative to the number redirects!

Maybe something like this:

```py
solr_docs
reading_log_work_keys

key_to_solr_doc = {doc[“key”]: doc for doc in solr_docs}

fetched_keys = set(doc[‘key’] for doc in solr_doc)
missing_keys = reading_log_work_keys - fetched_keys

new_docs = solr_get_many(missing_keys, ...)
Key_to_new_doc = {doc[“key”]: doc for doc in new_docs}

return [
   solr_docs.get(key) or new_docs.get(key)
    For key in reading_log_work_keys
]
```

File: openlibrary/core/bookshelves.py [hunk 14]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+                        scheme=WorkSearchScheme(),
+                        param={'q': '*:*'},
+                        offset=query_params["offset"],
+                        rows=limit,
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:275]: We actually don't need these params since we're fetching a specific number of works here instead. This might be the cause of the page 10 error above!

File: openlibrary/core/bookshelves.py [hunk 15]
```diff
@@ -230,6 +321,9 @@ def get_users_logged_books(
         from openlibrary.plugins.worksearch.code import run_solr_query
         from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
 
+        logger.info("BOOKSHELF: %s" % bookshelf_id)
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:324]: Don't miss these debug log lines!

File: openlibrary/core/bookshelves.py [hunk 16]
```diff
@@ -230,6 +321,9 @@ def get_users_logged_books(
...
+        # Sets the function to fetch editions as well, if not accessing the Want to Read shelf.
+        editions: bool = bookshelf_id != 1
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:326]: Rename variable to feel more bool-y

```suggestion
        show_editions: bool = bookshelf_id in != 1
```

File: openlibrary/plugins/upstream/mybooks.py [hunk 17]
```diff
@@ -19,7 +19,6 @@
 from openlibrary.core.models import LoggedBooksData
 from openlibrary.core.yearly_reading_goals import YearlyReadingGoals
 
-
```
Comment (cdrini) [inline openlibrary/plugins/upstream/mybooks.py:22]: Maybe drop these if you can since they're in an unrelated file.

**Review by cdrini — COMMENTED** [review]
Here's the finished review!

File: openlibrary/core/bookshelves.py [hunk 18]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+                if doc.get("logged_edition"):
+                    editions_to_work_doc.update({doc["logged_edition"]: doc})
+        # Attaches editions to the works, in second loop-- in case of misperformed order.
+        for edition in solr_docs:
+            if edition["key"].startswith("/books/"):
+                if work_doc := editions_to_work_doc.get(edition["key"]):
+                    work_doc.editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning(
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:225]: This should never happen, but logger.error will appear in sentry 

```suggestion
                    logger.error(
```

File: openlibrary/core/bookshelves.py [hunk 19]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+        # Ideally, this will be rare, as there's no way to access the relevant information through Solr.
+        for key in missing_keys.copy():
+            if not work_to_edition_keys.get(key):
+                missing_keys.remove(key)
+                solr_docs.append(web.storage({"key": key}))
+
+        keys_to_fetch = [
+            work_to_edition_keys[key].split("/")[2] for key in missing_keys
+        ]
+        if fq := f'edition_key:{" OR ".join(keys_to_fetch)}':
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:275]: This will be truthy regardless of `keys_to_fetch`

File: openlibrary/core/bookshelves.py [hunk 20]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+        """
+        Provides a proper 1-to-1 connection between work keys and edition keys; needed, in order to fill in the appropriate 'logged_edition' data
+        for the correctly pulled work later on, as well as to correctly update the post-redirect version back to the pre_redirect key.
+        Without this step, processes may error due to the 'post-redirect key' not actually existing within a user's reading log.
+        """
+        work_to_edition_keys = {
+            work: edition for (work, edition) in reading_log_keys if edition
+        }
+        edition_to_work_keys = {
+            edition.split("/")[2]: work
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:260]: It would be nice to keep the edition/work keys in the same format throughout the code ; we generally prefer the `/works/OL123W` format; so use that throughout, and only at the very end when construction the solr query change the format ; since it's only the solr `edition_key` field that's an oddball!
```suggestion
            edition: work
```

File: openlibrary/core/bookshelves.py [hunk 21]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+        }
+
+        # Here, we add in dummied works for situations in which there is no edition key present, yet the work key accesses a redirect.
+        # Ideally, this will be rare, as there's no way to access the relevant information through Solr.
+        for key in missing_keys.copy():
+            if not work_to_edition_keys.get(key):
+                missing_keys.remove(key)
+                solr_docs.append(web.storage({"key": key}))
+
+        keys_to_fetch = [
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:272]:
```suggestion
        edition_keys_to_query = [
```

File: openlibrary/core/bookshelves.py [hunk 22]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+        if fq := f'edition_key:{" OR ".join(keys_to_fetch)}':
+            solr_resp = run_solr_query(
+                scheme=WorkSearchScheme(),
+                param={'q': '*:*'},
+                fields=list(
+                    WorkSearchScheme.default_fetched_fields
+                    | {'subject', 'person', 'place', 'time', 'edition_key'}
+                ),
+                facet=False,
+                extra_params=[("fq", fq)],
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:284]: Add limit equal to the number of keys

File: openlibrary/core/bookshelves.py [hunk 23]
```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
...
+            ).docs
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:285]: Drop `docs` here so `solr_resp` below actually refers to a `SolrResponse` object

```suggestion
            )
```

File: openlibrary/core/bookshelves.py [hunk 24]
```diff
@@ -346,14 +421,28 @@ def get_filtered_reading_log_books(
...
                 # can be (surprisingly) slow if we have ~20k OR clauses.
-                extra_params=[('fq', f'key:({" OR ".join(reading_log_work_keys)})')],
+                extra_params=[('fq', filter_query)],
             )
             total_results = solr_resp.num_found
+            solr_docs = solr_resp.docs
+
+            if show_editions:
+                edition_data = get_solr().get_many(
+                    [f'{key}' for key in work_to_edition_keys.values()],
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:431]:
```suggestion
                    work_to_edition_keys.values(),
```

File: openlibrary/core/bookshelves.py (hunk 22 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py:284]: Done

File: openlibrary/core/bookshelves.py [hunk 25]
```diff
@@ -202,6 +203,104 @@ def count_total_books_logged_by_user_per_shelf(
...
+                ),
+                facet=False,
+                extra_params=[("fq", fq)],
+            )
+
+        """
+        Now, we add the correct 'logged_edition' information to each document retrieved by the query, and substitute the work_key in
+        each doc for the original one.
+        """
+        for doc in solr_resp.docs:
```
Comment (jimchamp) [inline openlibrary/core/bookshelves.py:292]: This fails when `edition_keys_to_query` is an empty list.  Example [here](https://testing.openlibrary.org/people/l_via_d_upinov_/books?debug=true).

Comment (benbdeitch) [reply openlibrary/core/bookshelves.py:293]: Thanks! I'm not sure how I let something this simple slip through. It was more or less an indent error originally, but it should be resolved now.

File: openlibrary/core/bookshelves.py [hunk 26]
```diff
@@ -332,12 +396,34 @@ def get_filtered_reading_log_books(
...
             assert len(reading_log_books) <= filter_book_limit
 
             # Wrap in quotes to avoid treating as regex. Only need this for fq
-            reading_log_work_keys = (
+            reading_log_work_keys = [
                 '"/works/OL%sW"' % i['work_id'] for i in reading_log_books
+            ]
+            reading_log_edition_keys = [
+                '"OL%sM"' % i['edition_id'] for i in reading_log_books
+            ]
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: This has been resolved in the latest versions.

File: openlibrary/core/bookshelves.py [hunk 27]
```diff
@@ -332,12 +396,34 @@ def get_filtered_reading_log_books(
...
+            logged_work_edition_dict = {
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: This was a great suggestion! The variable has been renamed in current versions.

File: openlibrary/core/bookshelves.py [hunk 28]
```diff
@@ -332,12 +396,34 @@ def get_filtered_reading_log_books(
...
+                '/works/OL%sW' % i['work_id']: '/books/OL%sM' % i['edition_id']
+                for i in reading_log_books
+            }
+            logged_work_edition_dict.update(
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Another fantastic suggestion.  Thanks!

File: openlibrary/core/bookshelves.py (hunk 2 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: This was extremely helpful, and made it into the final version.

File: openlibrary/core/bookshelves.py [hunk 29]
```diff
@@ -332,12 +396,34 @@ def get_filtered_reading_log_books(
...
+                    for i in reading_log_books
+                }
+            )
+
+            # Separating out the filter query from the call allows us to cleanly edit it, if editions are required.
+            filter_query = (
+                f'key:({" OR ".join(reading_log_work_keys)}) OR edition_key:({" OR ".join(reading_log_edition_keys)})'
+                if editions
+                else f'key:({" OR ".join(reading_log_work_keys)})'
             )
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Not bothering sounds good to me!

File: openlibrary/core/bookshelves.py (hunk 3 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: This was a great tip to learn, extremely useful, thanks!  I ended up having to change this to a list in the final version, so it ended up unused, however.

File: openlibrary/core/bookshelves.py (hunk 4 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Flatten has been used. The infogami implementation of it limits it to lists, so I was forced to change the tuple here to a list. Fortunately, that causes it to work just fine.

File: openlibrary/core/bookshelves.py (hunk 6 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: A simple, but useful elimination of an extraneous variable name. Thanks! Hopefully, these sorts of comments will be needed less over time.

File: openlibrary/core/bookshelves.py (hunk 7 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Another fantastic reduction of unnecessary variables! I appreciate it.  I'm learning that I haven't 'been using the 'startswith' method enough, either.

File: openlibrary/core/bookshelves.py (hunk 9 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: I feel like I have to put `docs_dict` in a picture frame somewhere, to remind myself of the sorts of variables that make perfect sense to me while coding, and completely nonsensical three days later. Thanks for the suggestion!

File: openlibrary/core/bookshelves.py (hunk 10 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Another wonderful adjustment in the world of passing pointers, rather than using saved indices to reference which docs go where; this is much appreciated, and I feel like it really helped the readability of this code.

File: openlibrary/core/bookshelves.py [hunk 30]
```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
...
+                    logger.warning("Error: No work found for edition %s" % full_key)
+        logger.info("LOGGERDICT IS %s" % docs_dict)
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
+        cls,
+        reading_log_work_keys,
+        solr_docs: list[web.Storage],
+        query_params,
```
[the remark this answers is not in the document]
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: It turns out that the offset was problematic, and contributing to a bug, so looks like we didn't need it even for that! Thanks again for the catch.

File: openlibrary/core/bookshelves.py (hunk 13 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: There's a lot of optimization that I'm still learning about; I was aware of O(n) notation, for example, but I wasn't sure when we particularly cared about it. The number of requests made is a perfect scenario for this, and I'm glad that the current version now packages all these requests in a batch.

File: openlibrary/core/bookshelves.py (hunk 11 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: As mentioned before, it's nice to remember that python makes such heavy use of references.  My previous version mapped keys to indices, but I'm now mapping keys to docs, as in this request.

File: openlibrary/core/bookshelves.py (hunk 15 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: They have been executed, as an example to the others. I suspect the debuggers will know better than to show up in future PRs.

File: openlibrary/core/bookshelves.py (hunk 14 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Indeed, the source of the error it was! Said error will never rear its ugly head again.

File: openlibrary/core/bookshelves.py (hunk 16 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: It has been thoroughly bool-ed, now. Thanks for all the advice with variable names, it shouldn't be as necessary in the future (I hope!)

File: openlibrary/plugins/upstream/mybooks.py (hunk 17 above)
Comment (benbdeitch) [reply openlibrary/plugins/upstream/mybooks.py]: Yeah, this is a relic of my old struggles with precommits. The unnecessary changes have been removed, and should not show up in the future.

**Comment by cdrini** [comment]
Performance-wise, so far this is looking good!

image

Will testing the filtered case once the bug is fixed causing it to return 0 results!

File: openlibrary/core/bookshelves.py (hunk 14 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Indeed, it was! The dreaded page 10 error has been eliminated, ne'er to rear its ugly head once more. (I hope)

File: openlibrary/core/bookshelves.py (hunk 20 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Change implemented.

File: openlibrary/core/bookshelves.py (hunk 24 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Change implemented, with an addendum. Instead of fetching all editions in the reading log now, we're only fetching the ones that match up to works.

File: openlibrary/core/bookshelves.py (hunk 23 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Fixed, thanks for the advice!

File: openlibrary/core/bookshelves.py [hunk 31]
```diff
@@ -202,6 +203,105 @@ def count_total_books_logged_by_user_per_shelf(
...
+        # Ideally, this will be rare, as there's no way to access the relevant information through Solr.
+        for key in missing_keys.copy():
+            if not work_to_edition_keys.get(key):
+                missing_keys.remove(key)
+                solr_docs.append(web.storage({"key": key}))
+
+        edition_keys_to_query = [
+            work_to_edition_keys[key].split("/")[2] for key in missing_keys
+        ]
+        fq = f'edition_key:{" OR ".join(edition_keys_to_query)}'
```
Comment (cdrini) [inline openlibrary/core/bookshelves.py:274]: Ahhh I think I found the issue!

```suggestion
        fq = f'edition_key:({" OR ".join(edition_keys_to_query)})'
```

Previously this output things like `edition_key:OL123M OR OL456M` , instead of `edition_key:(OL123M OR OL456M)`

Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: This shouldn't be the issue, but it's a definite oversight regardless. I've fixed it, and we'll see what happens.

File: openlibrary/core/bookshelves.py (hunk 19 above)
Comment (benbdeitch) [reply openlibrary/core/bookshelves.py]: Done

**Comment by cdrini** [comment]
Ok the bug was with ol-solr1 , so not related to this code! Performance is looking good as well: 
image

I'll do a few more tests later today, and then I think this is good to merge!!

**Review by cdrini — APPROVED** [review]
Ok lgtm! I tested redirects; there is _one_ edge case when you add two works and then the two works are merged... but I think it's very edge case and I'm not too sure how we should handle it. It _does_ work correctly when you add a work/edition and then that work gets merged into another. Also works if you add just a work, and it gets merged into another -- displays the empty tile it currently displays.
